src/main/java/org/codeforamerica/shiba/application/ApplicationRepository.java

```diff
@@ -140,6 +145,70 @@ private Timestamp convertToTimestamp(ZonedDateTime dateTime) {
         .orElse(null);
   }
 
+  public void updateStatusToInProgress(Application application,
+      RoutingDecisionService routingDecisionService) {
+    List<Document> documents = DocumentListParser.parse(application.getApplicationData());
+
+    for (Document document : documents) {
+      List<RoutingDestination> routingDestinations = routingDecisionService.getRoutingDestinations(
+          application.getApplicationData(), document);
+      Status status = application.getApplicationStatus(document);
+      if (status != DELIVERED) {
+        updateStatus(application.getId(), document, routingDestinations, IN_PROGRESS);
+      }
+    }
+  }
+
+  public void updateStatus(String id, Document document,
+      List<RoutingDestination> routingDestinations, Status status) {
+    routingDestinations.forEach(
+        routingDestination -> updateStatus(id, document, routingDestination.getName(), status));
+  }
+
+  public void updateStatus(String id, Document document, RoutingDestination routingDestination,
+      Status status) {
+    updateStatus(id, document, routingDestination.getName(), status);
+  }
+
+  /**
+   * Try to update existing status - if it's not found, add a new one.
+   */
+  public void updateStatus(String id, Document document, String routingDestination, Status status) {
+    updateStatus(id, document, status);
+    if (document == null || routingDestination == null) {
+      return;
+    }
+
+    String updateStatement = """
+        UPDATE application_status SET status = :status WHERE application_id = :application_id
+        AND document_type = :document_type AND routing_destination = :routing_destination
+        """;
+
+    Map<String, Object> parameters = new HashMap<>();
+    parameters.put("application_id", id);
+    parameters.put("status", status.toString());
+    parameters.put("document_type", document.name());
+    parameters.put("routing_destination", routingDestination);
+
+    var namedParameterJdbcTemplate = new NamedParameterJdbcTemplate(jdbcTemplate);
+
+    int rowCount = namedParameterJdbcTemplate.update(updateStatement, parameters);
+    if (rowCount == 0) {
+      // Not found, add a new entry
+      String insertStatement = """
+          INSERT INTO application_status (application_id, status, document_type, routing_destination)
+          VALUES (:application_id, :status, :document_type, :routing_destination)
+          """;
+      rowCount = namedParameterJdbcTemplate.update(insertStatement, parameters);
+    }
+
+    if (rowCount != 0) {
+      logStatusUpdate(id, document, routingDestination, status);
+    }
+
+  }
+
```

Comment:
I know it's already a big change but this feels like an opportunity for an `ApplicationStatusRepository` since this method doesn't actually write any SQL that affects the `applications` table.